File: app/src/main/java/com/github/mbarrben/moviedb/BaseActivity.java
package com.github.mbarrben.moviedb;

import android.app.Activity;
import android.os.Bundle;

import java.util.List;

import butterknife.ButterKnife;
import dagger.ObjectGraph;

public abstract class BaseActivity extends Activity {

    private ObjectGraph activityScopeGraph;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        injectDependencies();
        injectViews();
    }

    public void inject(Object object) {
        activityScopeGraph.inject(object);
    }

    protected abstract List<Object> getModules();

    private void injectDependencies() {
        MovieDbApp movieDbApp = (MovieDbApp) getApplication();
        List<Object> activityScopeModules = getModules();
        activityScopeModules.add(new ActivityModule(this));
        activityScopeGraph = movieDbApp.plus(activityScopeModules);
        inject(this);
    }

    private void injectViews() {
        ButterKnife.inject(this);
    }
}

File: app/src/main/kotlin/com/github/mbarrben/moviedb/BaseActivity.kt
package com.github.mbarrben.moviedb

import android.os.Bundle
import android.support.annotation.LayoutRes
import android.support.v7.app.AppCompatActivity
import com.github.mbarrben.moviedb.di.ActivityModule

abstract class BaseActivity(@LayoutRes val layoutRes: Int) : AppCompatActivity() {

  val applicationComponent by lazy { (application as MovieDbApp).applicationComponent }
  val activityModule by lazy { ActivityModule(this) }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(layoutRes)
    applicationComponent.inject(this)
    inject()
  }

  abstract fun inject()
}
